libPhoneNumber/NBPhoneNumberOfflineGeocoder.m

```diff
@@ -0,0 +1,173 @@
+  return (regionCode == NULL || [regionCode isEqualToString:@"ZZ"] ||
+          [regionCode isEqual:NB_REGION_CODE_FOR_NON_GEO_ENTITY])
+             ? @""
+             : [[[NSLocale alloc] initWithLocaleIdentifier:languageCode]
+                   displayNameForKey:NSLocaleCountryCode
+                               value:regionCode];
+}
+
+- (NSString *)descriptionForValidNumber:(NBPhoneNumber *)number
+                       withLanguageCode:(NSString *)languageCode {
+  [self->geocoderHelper setLanguage:languageCode];
+  NSString *descriptionResult = [self->geocoderHelper searchPhoneNumberInDatabase:number
+                                                                     withLanguage:languageCode];
+  if ([descriptionResult isEqualToString:@""]) {
+    return @"";
+  } else {
+    return descriptionResult;
+  }
+}
+
+- (NSString *)descriptionForValidNumber:(NBPhoneNumber *)phoneNumber
+                       withLanguageCode:(NSString *)languageCode
+                         withUserRegion:(NSString *)userRegion {
+  NSString *regionCode = [self->phoneUtil getRegionCodeForNumber:phoneNumber];
+  if ([userRegion isEqualToString:regionCode]) {
+    return [self descriptionForValidNumber:phoneNumber withLanguageCode:languageCode];
+  }
+
+  return [self regionDisplayName:regionCode withLanguageCode:languageCode];
+}
+
+- (NSString *)descriptionForNumber:(NBPhoneNumber *)phoneNumber
+                  withLanguageCode:(NSString *)languageCode {
+  NBEPhoneNumberType numberType = [self->phoneUtil getNumberType:phoneNumber];
+  if (numberType == NBEPhoneNumberTypeUNKNOWN) {
+    return @"";
+  } else if (![self->phoneUtil isNumberGeographical:phoneNumber]) {
+    return [self countryNameForNumber:phoneNumber withLanguageCode:languageCode];
+  }
+  return [self descriptionForValidNumber:phoneNumber withLanguageCode:languageCode];
+}
+
+- (NSString *)descriptionForNumber:(NBPhoneNumber *)phoneNumber
+                  withLanguageCode:(NSString *)languageCode
+                    withUserRegion:(NSString *)userRegion {
+  NBEPhoneNumberType numberType = [self->phoneUtil getNumberType:phoneNumber];
+  if (numberType == NBEPhoneNumberTypeUNKNOWN) {
+    return @"";
+  } else if (![self->phoneUtil isNumberGeographical:phoneNumber]) {
+    return [self countryNameForNumber:phoneNumber withLanguageCode:languageCode];
+  }
+  return [self descriptionForValidNumber:phoneNumber
+                        withLanguageCode:languageCode
+                          withUserRegion:userRegion];
+}
+
+- (NSString *)descriptionForNumber:(NBPhoneNumber *)phoneNumber {
+  NBEPhoneNumberType numberType = [self->phoneUtil getNumberType:phoneNumber];
+  NSString *languageCode;
+  // need to get from ios device settings
+  if ([[NSLocale preferredLanguages] count] > 0) {
+    // set language to preferred
+    languageCode = [NSLocale preferredLanguages][0];
+  } else {
+    languageCode = @"en";
+  }
```

Comment:
I think we should loop through the preferredLanguages list until we find nothing, then we return nil. We shouldn't assume that everyone is OK with English.

Reply:
Done.